### PR TITLE
refactor OpenMP `parallel_for` implementation to use more native OpenMP constructs

### LINKED ISSUE #4369: NGA FY22: refactor OpenMP backend to use more native OpenMP constructs
Some of the items to refactor/address: 
- [x] `parallel_for(RangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static)
- [x] `parallel_for(MDRangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static)
- [ ] `parallel_reduce(RangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static) 
  (follow OpenMPTarget backend for custom reductions)
- [ ] `parallel_reduce(MDRangePolicy)` using `#pragma omp parallel` for schedule(dynamic/static) 
  (follow OpenMPTarget backend for custom reductions)

### PR DESCRIPTION
**Pull request by cz4rs**
fixes #4369 

---
Initial benchmarking results: [link](https://github.com/NexGenAnalytics/kokkos/wiki/Benchmarks/)

### CODE REVIEW

File: BUILD.md [hunk 1]
```diff
@@ -177,6 +177,9 @@ Options can be enabled by specifying `-DKokkos_ENABLE_X`.
 * Kokkos_ENABLE_PROFILING_LOAD_PRINT
     * Whether to print information about which profiling tools gotloaded
     * BOOL Default: OFF
+* Kokkos_ENABLE_NATIVE_OPENMP
```
Comment (cz4rs) [inline BUILD.md:180]: This should be either `OFF` by default or (preferably) removed completely.

File: cmake/kokkos_enable_options.cmake [hunk 2]
```diff
@@ -55,6 +55,7 @@ KOKKOS_ENABLE_OPTION(COMPILER_WARNINGS    OFF "Whether to print all compiler war
 KOKKOS_ENABLE_OPTION(PROFILING_LOAD_PRINT OFF "Whether to print information about which profiling tools got loaded")
 KOKKOS_ENABLE_OPTION(TUNING               OFF "Whether to create bindings for tuning tools")
 KOKKOS_ENABLE_OPTION(AGGRESSIVE_VECTORIZATION OFF "Whether to aggressively vectorize loops")
+KOKKOS_ENABLE_OPTION(NATIVE_OPENMP        ON  "Whether to enable refactored OpenMP backend implementation")
```
Comment (cz4rs) [inline cmake/kokkos_enable_options.cmake:58]: Enabled temporarily to get the CI feedback.
This should be either `OFF` by default or (preferably) removed completely.

File: benchmarks/benchmark_suite/scripts/run_tests.bash [hunk 3]
```diff
@@ -3,15 +3,9 @@
 # BytesAndFlops
 cd build/bytes_and_flops
 
-USE_CUDA=`grep "_CUDA" KokkosCore_config.h | wc -l`
```
Comment (cz4rs) [inline benchmarks/benchmark_suite/scripts/run_tests.bash:6]: FIXME: revert this

**Comment by Rombur** [comment]
It seems that it's a bug in the compiler. Unfortunately, I cannot reproduce it outside of Kokkos. Here is what happens: we have a test  with a `for loop` that looks like `for(int i=0; i<0; ++i)` and for some reason when we use the dynamic scheduler, we enter the loop. Obviously that creates a segfault. The solution is to add 
```C++
     if (m_policy.begin() >= m_policy.end())
           return;
```
before the `#pragma omp parallel for` [here](https://github.com/kokkos/kokkos/pull/4664/files#diff-2918a925908a5355ef92cabbabc7f9e962815febb2a45bcc9ff9d428091fdf65R130). For me all the tests pass wit this change.

File: benchmarks/benchmark_suite/scripts/checkout_repos.bash [hunk 4]
```diff
@@ -3,11 +3,11 @@
 # Kokkos
 if [ ! -d "kokkos" ]; then
   git clone https://github.com/kokkos/kokkos
+  cd kokkos
+  git checkout develop
+  git pull
+  cd ..
 fi
```
Comment (dalg24) [inline benchmarks/benchmark_suite/scripts/checkout_repos.bash:10]: Why did you change?  The intent seemed to be to pull the latest version when the repo had already been cloned

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 5]
```diff
@@ -56,6 +56,20 @@
...
+#undef KOKKOS_PRAGMA_IVDEP_IF_ENABLED
+#define KOKKOS_PRAGMA_IVDEP_IF_ENABLED _Pragma(ivdep)
+#endif
+#endif
+
+#ifndef KOKKOS_COMPILER_NVHPC
+#define KOKKOS_OPENMP_OPTIONAL_CHUNK_SIZE , m_policy.chunk_size()
+#else
+#define KOKKOS_OPENMP_OPTIONAL_CHUNK_SIZE
+#endif
```
Comment (dalg24) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:71]: Make sure you `#undef` them

File: benchmarks/benchmark_suite/scripts/checkout_repos.bash (hunk 4 above)
Comment (cz4rs) [reply benchmarks/benchmark_suite/scripts/checkout_repos.bash:10]: True, that's a change of behavior, but I thought it would be more convenient. You don't need to modify the script when you are running the benchmark on a branch and no surprising pulls happen.

Comment (cz4rs) [reply benchmarks/benchmark_suite/scripts/checkout_repos.bash:10]: That's a bit different from the other libraries in the script, but I think it should be fine since it's Kokkos that we are benchmarking here.

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 6]
```diff
@@ -92,53 +102,92 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
...
+      exec_work(const FunctorType& functor, const Member iwork) {
+    const TagType t{};
+    functor(t, iwork);
+  }
+
+  template <class Policy>
+  typename std::enable_if<std::is_same<typename Policy::schedule_type::type,
+                                       Kokkos::Dynamic>::value>::type
+  execute_parallel() const {
+    if (m_policy.begin() >= m_policy.end()) return;
```
Comment (Rombur) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:130]: Add a comment that we do this because we hit a compiler bug with nvhpc 11.4.

**Review by Rombur — APPROVED** [review]

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 7]
```diff
@@ -92,53 +102,93 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
...
-    for (Member iwork = ibeg; iwork < iend; ++iwork) {
+    KOKKOS_PRAGMA_IVDEP_IF_ENABLED
+    for (auto iwork = ibeg; iwork < iend; ++iwork) {
       functor(t, iwork);
     }
   }
 
+  template <class TagType>
+  inline static
+      typename std::enable_if<std::is_same<TagType, void>::value>::type
```
Comment (dalg24) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:113]:
```suggestion
  template <class Enable = WorkTag>
  static std::enable_if_t<std::is_void<WorkTag>::value && std::is_same<Enable, WorkTag>::value>
```
Then at call site `ParallelFor::template exec_work<WorkTag>(m_functor, iWork) -> exec_work(m_functor, iWork);`

Comment (dalg24) [reply core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp]: Yes, not a bad idea to fix that one too

**Comment by Rombur** [comment]
Can we merge this. I have PRs that are almost ready but they conflict with this PR and I am waiting for this code to be merged to finish them.

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 8]
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
...
+  inline static typename std::enable_if_t<!std::is_void<WorkTag>::value &&
+                                          std::is_same<Enable, WorkTag>::value>
+  exec_work(const FunctorType& functor, const Member iwork) {
+    const Enable t{};
+    functor(t, iwork);
+  }
+
+  template <class Policy>
+  typename std::enable_if<std::is_same<typename Policy::schedule_type::type,
+                                       Kokkos::Dynamic>::value>::type
```
Comment (masterleinad) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:126]:
```suggestion
  std::enable_if_t<std::is_same<typename Policy::schedule_type::type, Kokkos::Dynamic>::value>
```
?

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 9]
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
...
+    const Enable t{};
+    KOKKOS_PRAGMA_IVDEP_IF_ENABLED
+    for (auto iwork = ibeg; iwork < iend; ++iwork) {
       functor(t, iwork);
     }
   }
 
+  template <class Enable = WorkTag>
+  inline static typename std::enable_if_t<std::is_void<WorkTag>::value &&
+                                          std::is_same<Enable, WorkTag>::value>
```
Comment (masterleinad) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:111]:
```suggestion
  inline static std::enable_if_t<std::is_void<WorkTag>::value && std::is_same<Enable, WorkTag>::value>
```
?

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 10]
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
...
+  exec_work(const FunctorType& functor, const Member iwork) {
+    functor(iwork);
+  }
+
+  template <class Enable = WorkTag>
+  inline static typename std::enable_if_t<!std::is_void<WorkTag>::value &&
+                                          std::is_same<Enable, WorkTag>::value>
```
Comment (masterleinad) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:118]:
```suggestion
  inline static std::enable_if_t<!std::is_void<WorkTag>::value && std::is_same<Enable, WorkTag>::value>
```
?

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 11]
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
...
-    const TagType t{};
-#ifdef KOKKOS_ENABLE_AGGRESSIVE_VECTORIZATION
-#ifdef KOKKOS_ENABLE_PRAGMA_IVDEP
-#pragma ivdep
-#endif
-#endif
-    for (Member iwork = ibeg; iwork < iend; ++iwork) {
+  template <class Enable = WorkTag>
+  inline static typename std::enable_if_t<!std::is_void<WorkTag>::value &&
+                                          std::is_same<Enable, WorkTag>::value>
```
Comment (masterleinad) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:100]:
```suggestion
  inline static std::enable_if_t<!std::is_void<WorkTag>::value && std::is_same<Enable, WorkTag>::value>
```
?

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 12]
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
...
-      const FunctorType& functor, const Member ibeg, const Member iend) {
-#ifdef KOKKOS_ENABLE_AGGRESSIVE_VECTORIZATION
-#ifdef KOKKOS_ENABLE_PRAGMA_IVDEP
-#pragma ivdep
-#endif
-#endif
-    for (Member iwork = ibeg; iwork < iend; ++iwork) {
+  template <class Enable = WorkTag>
+  inline static typename std::enable_if_t<std::is_void<WorkTag>::value &&
+                                          std::is_same<Enable, WorkTag>::value>
```
Comment (masterleinad) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:90]:
```suggestion
  inline static std::enable_if_t<std::is_void<WorkTag>::value && std::is_same<Enable, WorkTag>::value>
```
?

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 13]
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
...
-    for (Member iwork = ibeg; iwork < iend; ++iwork) {
+  template <class Enable = WorkTag>
+  inline static typename std::enable_if_t<!std::is_void<WorkTag>::value &&
+                                          std::is_same<Enable, WorkTag>::value>
+  exec_range(const FunctorType& functor, const Member ibeg, const Member iend) {
+    const Enable t{};
+    KOKKOS_PRAGMA_IVDEP_IF_ENABLED
+    for (auto iwork = ibeg; iwork < iend; ++iwork) {
       functor(t, iwork);
     }
```
Comment (masterleinad) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:106]:
```suggestion
    const WorkTag tag{};
    KOKKOS_PRAGMA_IVDEP_IF_ENABLED
    for (auto iwork = ibeg; iwork < iend; ++iwork) {
      functor(tag, iwork);
    }
```

File: benchmarks/benchmark_suite/scripts/run_benchmark.bash [hunk 14]
```diff
@@ -1,14 +1,33 @@
...
+readonly COMPILER=$4
+if [[ $# -lt 4 ]]; then
+  echo "Usage: ./run_benchmark.bash PATH_TO_SCRIPTS KOKKOS_DEVICES KOKKOS_ARCH COMPILER"; exit 1
+fi
+
+function get_logfile() {
+  local logfile=run-benchmark--$(date +"%Y-%m-%d-%H-%M").log
+
+  cd ${PWD}/kokkos
+  local git_info="KOKKOS_BRANCH: $(git branch --show-current)
```
Comment (cz4rs) [inline benchmarks/benchmark_suite/scripts/run_benchmark.bash:14]: `git branch --show-current` requires Git 2.22. I can use `rev-parse` here if that's too fresh for some reason.

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 15]
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
...
+    num_threads(OpenMP::impl_thread_pool_size())
+    KOKKOS_PRAGMA_IVDEP_IF_ENABLED
+    for (auto iwork = m_policy.begin(); iwork < m_policy.end(); ++iwork) {
+      ParallelFor::template exec_work(m_functor, iwork);
+    }
+  }
+
+  template <class Policy>
+  typename std::enable_if<!std::is_same<typename Policy::schedule_type::type,
+                                        Kokkos::Dynamic>::value>::type
```
Comment (masterleinad) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:140]:
```suggestion
  std::enable_if_t<!std::is_same<typename Policy::schedule_type::type, Kokkos::Dynamic>::value>
```

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 16]
```diff
@@ -71,70 +85,108 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
...
+    }
+
+    OpenMPExec::verify_is_master("Kokkos::OpenMP parallel_for");
+#ifndef KOKKOS_INTERNAL_DISABLE_NATIVE_OPENMP
+    execute_parallel<Policy>();
+#else
     enum {
       is_dynamic = std::is_same<typename Policy::schedule_type::type,
                                 Kokkos::Dynamic>::value
     };
```
Comment (masterleinad) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:164]:
```suggestion
    constexpr bool is_dynamic = std::is_same<typename Policy::schedule_type::type, Kokkos::Dynamic>::value;
```

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 17]
```diff
@@ -167,55 +219,84 @@ class ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>,
...
```
Comment (masterleinad) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:272]:
```suggestion
    constexpr bool is_dynamic = std::is_same<typename Policy::schedule_type::type, Kokkos::Dynamic>::value;
```

**Review by masterleinad — COMMENTED** [review]
I would prefer if all the changes to the `benchmark` directory would be in a separate pull request.

**Comment by PhilMiller** [comment]
retest this please

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 18]
```diff
@@ -167,55 +216,83 @@ class ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>,
...
     if (OpenMP::in_parallel()) {
       ParallelFor::exec_range(m_mdr_policy, m_functor, m_policy.begin(),
                               m_policy.end());
-    } else {
-      OpenMPExec::verify_is_master("Kokkos::OpenMP parallel_for");
+      return;
+    }
+
+    OpenMPExec::verify_is_master("Kokkos::OpenMP parallel_for");
+#ifndef KOKKOS_INTERNAL_DISABLE_NATIVE_OPENMP
```
Comment (masterleinad) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:263]: Where is this defined?

Comment (cz4rs) [reply core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:263]: This is not defined anywhere - it's intended as an advanced option for developers (so you'd have to go with `add_definitions(-DKOKKOS_INTERNAL_DISABLE_NATIVE_OPENMP)` or similar). This was requested not to be exposed as public option, but a temporary fallback.

Comment (masterleinad) [reply core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:263]: Ah, OK.

**Review by masterleinad — APPROVED** [review]

**Comment by cz4rs** [comment]
Rebased on top of `develop`, hopefully jenkins likes it this time.

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 19]
```diff
@@ -71,70 +84,106 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
...
-    for (Member iwork = ibeg; iwork < iend; ++iwork) {
+  template <class Enable = WorkTag>
+  inline static std::enable_if_t<std::is_void<WorkTag>::value &&
+                                 std::is_same<Enable, WorkTag>::value>
+  exec_range(const FunctorType& functor, const Member ibeg, const Member iend) {
+    KOKKOS_PRAGMA_IVDEP_IF_ENABLED
+    for (auto iwork = ibeg; iwork < iend; ++iwork) {
       functor(iwork);
     }
   }
```
Comment (dalg24) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:95]: Why is not just
```suggestion
  inline static exec_range(const FunctorType& functor, const Member ibeg, const Member iend) {
    KOKKOS_PRAGMA_IVDEP_IF_ENABLED
    for (auto iwork = ibeg; iwork < iend; ++iwork) {
      exec_work(functor, iwork);
    }
  }
```

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 20]
```diff
@@ -167,55 +216,83 @@ class ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>,
...
 
+  inline static void exec_work(const MDRangePolicy& mdr_policy,
+                               const FunctorType& functor, const Member iwork) {
+    iterate_type(mdr_policy, functor)(iwork);
+  }
+
+  template <class Policy>
+  typename std::enable_if<std::is_same<typename Policy::schedule_type::type,
+                                       Kokkos::Dynamic>::value>::type
```
Comment (dalg24) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:232]: Please use `enable_if_t`

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp (hunk 19 above)
Comment (cz4rs) [reply core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp]: Added that and simplified the implementation for `MDRangePolicy` as well :+1:

**Comment by cz4rs** [comment]
@dalg24 I have addressed all the comments.

Jenkins has two builds failing:
- `OpenMPTarget` (`KokkosCore_UnitTest_OpenMPTarget`)
- `SYCL-OneApi` (`KokkosCore_PerfTestExec`)
I believe these are unrelated to this PR. If so, hopefully this can be merged.

**Review by dalg24 — APPROVED** [review]
